storage_service, group0_state_machine: move SL cache update from `topology_state_load()` to `load_snapshot()`

Jadw1 opened the pull request:
Currently the service levels cache is unnecessarily updated in every call of `topology_state_load()`.
But it is enough to reload it only when a snapshot is loaded. 
(The cache is also already updated when there is a change to one of `service_levels_v2`, `role_members`, `role_attributes` tables.)

Fixes scylladb/scylladb#25114
Fixes scylladb/scylladb#23065

All versions with service levels on raft has this issue, so this patch can be backported to fix it.

Jadw1 commented:
@mykaul @patjed41 I'm not sure whether we should backport this PR or not?

patjed41 reviewed (approved):
This PR should also fix the initial issue - #23065. Based on the logs, the timeout happened in `topology_transition -> topology_state_load -> update_service_levels_cache`, not `load_snapshot -> topology_state_load -> update_service_levels_cache`.

Even thought the issue will be fixed, I think we should still do something about this timeout. Maybe we can fix 53224d90bea6629fcb8369706daee4dc1c0504cb, which didn't pass `ctx` to `update_effective_service_levels_cache()`.

cc @emdroid

piodul commented on service/storage_service.cc:
```diff
@@ -900,7 +899,10 @@ future<> storage_service::merge_topology_snapshot(raft_snapshot snp) {
 future<> storage_service::update_service_levels_cache(qos::update_both_cache_levels update_only_effective_cache, qos::query_context ctx) {
     SCYLLA_ASSERT(this_shard_id() == 0);
-    co_await _sl_controller.local().update_cache(update_only_effective_cache, ctx);
+    if (_topology_state_machine._topology.upgrade_state == topology::upgrade_state_type::done) {
+        // Skip cache update unless the topology upgrade is done
+        co_await _sl_controller.local().update_cache(update_only_effective_cache, ctx);
+    }
```
Would checking the service level controller version instead of the topology upgrade state make more sense here?

Jadw1 commented:
v3 changelog:
- addressed comment: switched `_topology_state_machine._topology.upgrade_state == topology::upgrade_state_type::done` to `_sl_controller.local().is_v2()`
